feat: implement --shdict to create shared dict

thibaultcha opened the pull request:
Updated version of #11, as a new PR because of the semantic changes.

This implements an `--shdict` option, as previously discussed. It leaves the room open to including other ngx_lua directives (including ones with potentially multiple occurrences).

thibaultcha commented:
(I will update the docs if this is fine)

thibaultcha commented:
@agentzh Did you get a chance to look at this implementation?

agentzh commented:
@thibaultcha Sorry for the delay on my side. Your patch looks good. Will you update the docs accordingly? Many thanks!

thibaultcha commented:
Patch updated! 

* updated the `--help` message in README
* added `--shdict` example in README

Could you also take a look at #13? Both of those features would allow avoiding various hacks when scripting with resty-cli :wink:

thibaultcha commented:
@agentzh I have rebased this on master. Would you take a look to it as well? Thanks!

agentzh commented on README.md:
```diff
@@ -56,6 +56,9 @@ Synopsis
 
         --resolve-ipv6      Make the nginx resolver lookup both IPv4 and IPv6 addresses
 
+        --shdict            Create the specified lua shared dicts in the http
```
Let's be explicit about this option's argument: `--shdict NAME`

agentzh commented on bin/resty:
```diff
@@ -419,6 +430,9 @@ Options:
 
     --resolve-ipv6      Make the nginx resolver lookup both IPv4 and IPv6 addresses
 
+    --shdict            Create the specified lua shared dicts in the http
```
Ditto.

agentzh commented on README.md:
```diff
@@ -56,6 +56,9 @@ Synopsis
 
         --resolve-ipv6      Make the nginx resolver lookup both IPv4 and IPv6 addresses
 
+        --shdict NAME       Create the specified lua shared dicts in the http
```
Sorry, I've had a closer look and found that the real syntax is actually `--shdict "NAME SIZE"` here?

agentzh commented on bin/resty:
```diff
@@ -213,6 +222,7 @@ for my $var (sort keys %ENV) {
 
 my $main_include_directives = resolve_includes('main', @main_includes);
 my $http_include_directives = resolve_includes('http', @http_includes);
+my $lua_shared_dicts = format_directives('lua_shared_dict', @shdicts);
```
I think it's better to check the syntax of the shdict option values here since error messages from nginx might not be very friendly and helpful :)

thibaultcha replied:
Makes sense. What would be a proper error message and behavior? Something like this maybe?
```
Invalid value for '--shdict': <...>
```

Or followed by the usage instructions?
```
Invalid value for '--shdict': <...>

resty [options] [lua-file [args]]

Options:
    -c num              Set maximal connection count (default: 64).
    -e prog             Run the inlined Lua code in "prog".
    --help              Print this help.

    --http-include path Include the specified file in the nginx http configuration block
                        (multiple instances are supported).

    -I dir              Add dir to the search paths for Lua libraries.

    --main-include path Include the specified file in the nginx main configuration block
                        (multiple instances are supported).

    --nginx             Specify the nginx path (this option might be removed in the future).
    -V                  Print version numbers and nginx configurations.
    --valgrind          Use valgrind to run nginx
    --valgrind-opts     Pass extra options to valgrind

For bug reporting instructions, please see:

    <https://openresty.org/en/community.html>

Copyright (C) Yichun Zhang (agentzh). All rights reserved.
```

Both options returning an exit code of `2`? Or `1`?

thibaultcha replied:
I'd rather ask before instead of implementing something and having to iterate over it again because you had something else in mind :)

thibaultcha replied:
@agentzh Let me know what you think of 356904e20c641c0eb10d19262089386ef50a45a8

agentzh commented on bin/resty:
```diff
@@ -47,6 +66,8 @@ GetOptions("c=i",             \(my $conns_num),
            "V|v",             \(my $version))
    or usage(1);
 
+check_arg_format('--shdict', qr/^[a-z]*\s\d+(?i)[km]$/, 'NAME SIZE', @shdicts);
```
This regex looks wrong to me. Some notes:

1. The shared dict name is mandatory while `*` allows empty names.
2. The shared dict name can contain underscores.
3. The size units `k` and `m` are optional. default to bytes.
4. we should allow more than more spaces between `NAME` and `SIZE`.
5. `\z` is better than `$` here to avoid multiple lines in the value.

Also, it's better to pass the `@shdicts` array as an array reference, that is, `\@shdicts`.

thibaultcha replied:
Yes, this regex was totally overlooked, my bad.

> The size units k and m are optional. default to bytes.

I was aware of this according to the nginx docs but somehow it seems not to be the case?

```
resty -e "" --shdict="dogs 1024"
nginx: [emerg] invalid lua shared dict size "1024" in /var/folders/g0/yjhyxj6d4nz4vdfs2hvwy5l80000gn/T/t1XZsHoN75/conf/nginx.conf:53
```

Also: should we allow capital letters in the shdict name as well?

agentzh replied:
@thibaultcha Try larger values? I remember there is a minimum size restriction like 8KB or 12KB.

agentzh commented on bin/resty:
```diff
@@ -30,14 +30,33 @@ sub resolve_includes {
+    my ($arg_name, $regex, $arg_format, @args) = @_;
+    for (@args) {
+        unless (m/$regex/) {
+            warn "Invalid value for $arg_name option. Expected: $arg_format\n";
```
We should have used `die` here directly.

thibaultcha commented:
@agentzh I have updated the regex in cf1d137.

thibaultcha commented:
@agentzh Do you have any other concerns regarding the new regex? Thanks!

agentzh commented:
@thibaultcha Sorry for the delay on my side. It looks good to me.

@doujiang24 @membphis Will you have a quick look at this PR too? Thanks!

doujiang24 commented:
@thibaultcha Nice job 👍 
@agentzh LGTM:)

agentzh commented:
@thibaultcha Merged with slightly modified changes. Thanks!
